### PR TITLE
Features to pass method not found feature in acceptance tests

### PR DESCRIPTION
**Pull request by ikarabulut**
**Acceptance Test that is now passing** -> https://github.com/8thlight/http_server_spec/blob/master/features/01_getting_started/method_not_allowed.feature

**Changes Include:**
- The responsibility to know what paths are accepted has been moved from the `RequestRunner` classes to the `Router`. Now when a `Router` is constructed, an instance variable is created with the K/V pairings of a path to its accepted Methods. The Router then compares the Method call to the corresponding Method values in the Paths map. If the values match up then the corresponding `RequestRunner` Method is executed.
- If the Method does not match up to the accepted Methods for that specified route then the client will be sent to `notAllowed()` which takes in the accepted paths for that path, and then executes the `MethodNotAllowedRequestHandler`

This leads to the next change and coming refactoring: _I will be updating the names of the `RequestRunners` to be `RequestHandlers` they will no longer know their valid paths as that responsibility has moved to the `Route` class. Also my plan is to be able to dynamically generate headers within the `RequestHandler` similar to how it is now in `MethodNotAllowedRequestHandler` except my plan is to create a new class which knows available headers and can generate the correct K/V pairing_

**Coming Changes once this is reviewed**
- I will be encapsulating the `Request` into its own object which will be passed instead of just the `initialLine` map. 
- The `requests` directory will be updated to `handlers` which will handle specific request types, and will implement a `Handler` interface
- A new class **or** directory will be created which handles headers and knows what headers are available and how to generate them with correct metadata

I think that covers everything. There are a few syntax refactorings as well. I ended up updating a few tests for the `RouteTest` since we need to pass in valid paths (for now) or else the test returns a `null pointer`. Also `HeadResponse` had an additional `CRLF` in `stringifyResponse` which was removed

### CODE REVIEW

**Comment by jramnani** [comment]
> There are a few syntax refactorings as well. I ended up updating a few tests for the RouteTest since we need to pass in valid paths (for now) or else the test returns a null pointer. Also HeadResponse had an additional CRLF in stringifyResponse which was removed

Are there tests for these conditions?

File: app/src/main/java/com/ikarabulut/response/StatusCode.java [hunk 1]
```diff
@@ -1,16 +1,20 @@
 package com.ikarabulut.response;
 
 public enum StatusCode {
-    OK("200");
+    OK("OK" , "200"), NOT_ALLOWED("NOT ALLOWED", "405");
```
Comment (jramnani) [inline app/src/main/java/com/ikarabulut/response/StatusCode.java:4]: Suggestion: I think Enum values are put on separate lines. That way they can be easier to scan, especially if they are ordered in some way, like by their return code.

File: app/src/main/java/com/ikarabulut/server/Router.java [hunk 2]
```diff
@@ -2,24 +2,43 @@
...
+    private Map<String, List<String>> paths;
 
     public Router(Map<String, String> initialLine) {
         this.initialLine = initialLine;
+        generateRoutes();
+    }
+
+    public void generateRoutes() {
+        this.paths = new HashMap<>() {{
+            put("/head_request", new ArrayList<>(Arrays.asList("HEAD", "OPTIONS")));
```
Comment (jramnani) [inline app/src/main/java/com/ikarabulut/server/Router.java:22]: Question: Do you need to say `new ArrayList` here?  Or can you use the return value of `Arrays.asList` as the value?

File: app/src/main/java/com/ikarabulut/server/Router.java [hunk 3]
```diff
@@ -2,24 +2,43 @@
...
     public Response routeRequest() {
         String method = initialLine.get("httpMethod");
-        switch (method) {
-            case "HEAD":
-                return runHeadRequest();
-            case "GET":
-                return runGetRequest();
+        String path = initialLine.get("httpPath");
+
+        if (paths.get(path).contains(method)) {
```
Comment (jramnani) [inline app/src/main/java/com/ikarabulut/server/Router.java:32]: 👍

Comment (jramnani) [inline app/src/main/java/com/ikarabulut/server/Router.java:32]: Here is an example where an using the "return early" technique would be preferred.

Why? Because the way the code is currently written it nests conditionals. We are at two levels deep of nesting, but if a special case develops for one of those methods then the natural thing to do would be add _another_ layer of nesting.

```
if (foo) {
  switch (method) {
    case "HEAD":
      if request.path == "/special" {
         doSomethingSpecial();
      } else {
        return runHeadRequest();
      }
   } else {
     importantErrorHandling();
  }
}
```

This is how previously reasonable code turns into hard-to-follow spaghetti.

By returning early we do data validation early, return early if data is invalid, then the rest of the method can continue knowing that the data is safe (it is not null, it is a valid HTTP method name, etc.)

```
if !(paths.get(path.contains(method)) {
  // return early
  List<String> allowedMethods = paths.get(path);
  return notAllowed(paths.get(allowedMethods));
}

switch (method) {
  case "HEAD":
    return runHeadRequest();
  case "GET":
    return runGetRequest();
}
```

Notice how this "pulls the code left" and reduces nesting.  (Nesting is a proxy for conditionals.)  This code is easier to read and follow what is going on.

The trade-off here is that we are doing a negated boolean check in the `if` statement which is a little harder to read. But that can be fixed with a well-named helper method. e.g. `if (!isMethodAllowedForPath(path)) { ... }`

File: app/src/test/java/com/ikarabulut/response/HeadResponseTest.java [hunk 4]
```diff
@@ -26,7 +26,7 @@ void stringifyResponse_ForHead() {
         String expectedResponse = version + " " + statusNumber + " " + statusCode + "\r\n" +
                 "Date: " + headers.get("Date") + "\r\n" +
                 "Content-Language: " + headers.get("Content-Language") + "\r\n" +
-                "\r\n" + "\r\n";
+                "\r\n";
```
Comment (jramnani) [inline app/src/test/java/com/ikarabulut/response/HeadResponseTest.java:29]: Oh, I see what the issue is, there was already a `\r\n` on the previous line, so the test had 3 separators intead of two.

**Review by jramnani — APPROVED** [review]
Some questions and suggestions, but no blockers. ✅

File: app/src/main/java/com/ikarabulut/server/Router.java (hunk 3 above)
Comment (ikarabulut) [reply app/src/main/java/com/ikarabulut/server/Router.java:32]: New concept! sweet thank you.

I updated the implementation of `routeRequest()` and extracted a few private methods which makes it easier to read

```
    public Response routeRequest() {

        if (!pathIncludesMethod()) {
            return methodNotAllowed(pathsIncludedMethods());
        }

        switch (initialLine.get("httpMethod")) {
            case "HEAD":
                return runHeadRequest();
            case "GET":
                return runGetRequest();
        }
        return null;
    }
```

**Comment by ikarabulut** [comment]
Applied some minor tweaks based on the feedback. One additional change outside of the feedback was I updated `notAllowed()` to `methodNotAllowed()` so its represents what it actually means. Merging to main
